lib/timers.js

```diff
@@ -164,56 +196,82 @@ exports._unrefActive = function(item) {
 // Appends a timer onto the end of an existing timers list, or creates a new
 // TimerWrap backed list if one does not already exist for the specified timeout
 // duration.
-function insert(item, unrefed, start) {
+function insert(item, refed, start) {
   const msecs = item._idleTimeout;
-  if (msecs < 0 || msecs === undefined) return;
-
-  if (typeof start === 'number') {
-    item._idleStart = start;
-  } else {
-    item._idleStart = TimerWrap.now();
-  }
+  if (msecs < 0 || msecs === undefined)
+    return;
 
-  const lists = unrefed === true ? unrefedLists : refedLists;
+  item._idleStart = start;
 
   // Use an existing list if there is one, otherwise we need to make a new one.
   var list = lists[msecs];
   if (list === undefined) {
     debug('no %d list was found in insert, creating a new one', msecs);
-    lists[msecs] = list = new TimersList(msecs, unrefed);
+    lists[msecs] = list = new TimersList(start, msecs);
+    queue.insert(list);
+
+    if (nextExpiry > start + msecs) {
+      if (handle === null)
+        createHandle(refed);
+      handle.start(msecs);
+      nextExpiry = start + msecs;
+    }
   }
 
   if (!item[async_id_symbol] || item._destroyed) {
     item._destroyed = false;
     initAsyncResource(item, 'Timeout');
   }
 
+  if (refed === !item[kRefed]) {
+    if (refed)
+      incRefCount();
+    else
+      decRefCount();
+  }
+  item[kRefed] = refed;
+
   L.append(list, item);
-  assert(!L.isEmpty(list)); // list is not empty
 }
 
-function TimersList(msecs, unrefed) {
+function TimersList(start, msecs) {
   this._idleNext = this; // Create the list with the linkedlist properties to
   this._idlePrev = this; // prevent any unnecessary hidden class changes.
-  this._unrefed = unrefed;
+  this.expiry = start + msecs;
+  this.id = timerListId++;
   this.msecs = msecs;
-
-  const timer = this._timer = new TimerWrap();
-  timer._list = this;
-
-  if (unrefed === true)
-    timer.unref();
-  timer.start(msecs);
+  this.priorityQueuePosition = null;
 }
 
+const { _tickCallback: runNextTicks } = process;
 function processTimers(now) {
-  if (this.owner)
-    return unrefdHandle(this.owner, now);
-  return listOnTimeout(this, now);
+  debug('process timer lists %d', now);
+  nextExpiry = Infinity;
+
+  let list, ran;
+  while (list = queue.peek()) {
+    if (list.expiry > now)
+      break;
+    if (ran)
+      runNextTicks();
```

Comment:
This won't cause problems if you schedule short timers in next ticks and the existing queues take longer than that amount of time to process?

This is also probably worth leaving a comment for since we usually don't manually run nextTicks.

Reply:
Sure but we can just come back on the next iteration of the event loop. The current behaviour is pretty problematic because it's possible to just get stuck in timers literally looping forever if nextTicks take too long.

Reply:
Currently timers get called repeatedly but the event loop still turns as far as I'm aware. Here you might loop forever in JS, no?

Reply:
Current behaviour is... complicated but basically if you run a list, then add a timeout in next tick, then that next tick takes a long time to follow up, then that list will run again, and then you could keep doing all that over and over without a stop.

With the new behaviour, when we call in to execute timers, we only run all timers that were due at that exact time. The concept of `now` never updates. So any timers that weren't due during that loop's tick will wait until the next one.

Reply:
> Current behaviour is...

Hmmm, I thought we had fixed something like that but perhaps not _that_.

> With the new behaviour, ... The concept of `now` never updates

Ehhh, I just hope that doesn't cause any userland ordering problems.